src/wait.rs

```diff
@@ -0,0 +1,288 @@
+    /// # #[cfg(feature = "rustls-tls")]
+    /// # let mut client = ClientBuilder::rustls().connect("http://localhost:4444").await.expect("failed to connect to WebDriver");
+    /// # #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+    /// # let mut client: fantoccini::Client = unreachable!("no tls provider available");
+    /// // -- snip wrapper code --
+    /// let button = client.wait().on(Locator::Css(
+    ///     r#"a.button-download[href="/learn/get-started"]"#,
+    /// )).await?;
+    /// // -- snip wrapper code --
+    /// # client.close().await
+    /// # }
+    /// ```
+    pub fn new(client: &'c mut Client) -> Self {
+        Self {
+            client,
+            timeout: Some(DEFAULT_TIMEOUT),
+            period: DEFAULT_PERIOD,
+        }
+    }
+
+    /// Set the timeout until the operation should wait.
+    pub fn at_most(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    /// Wait forever.
+    pub fn forever(mut self) -> Self {
+        self.timeout = None;
+        self
+    }
+
+    /// Sets the period to delay checks.
+    pub fn every(mut self, period: Duration) -> Self {
+        self.period = period;
+        self
+    }
+
+    /// Wait until a condition exists or a timeout is hit.
+    pub async fn on_condition<T, F>(self, f: F) -> Result<T, error::CmdError>
+    where
+        F: for<'f> FnMut(
+            &'f mut Client,
+        ) -> Pin<
+            Box<dyn Future<Output = Result<Option<T>, error::CmdError>> + 'f + Send>,
+        >,
+    {
+        self.on(Condition(f)).await
+    }
+
+    /// Wait until a condition exists or a timeout is hit.
+    pub async fn on<C, T>(self, mut condition: C) -> Result<T, error::CmdError>
+    where
+        C: for<'a> WaitCondition<'a, T>,
+    {
+        let start = Instant::now();
+        loop {
+            match self.timeout {
+                Some(timeout) if start.elapsed() > timeout => {
+                    break Err(error::CmdError::WaitTimeout)
+                }
+                _ => {}
+            }
+            match condition.ready(self.client).await? {
+                Some(result) => break Ok(result),
+                None => {
+                    tokio::time::sleep(self.period).await;
+                }
+            };
+        }
+    }
+
+    /// Wait for a predicate.
+    pub async fn on_predicate<F>(self, predicate: F) -> Result<(), error::CmdError>
+    where
+        F: for<'f> FnMut(
+            &'f mut Client,
+        ) -> Pin<
+            Box<dyn Future<Output = Result<bool, error::CmdError>> + 'f + Send>,
+        >,
+    {
+        self.on(Predicate(predicate)).await
+    }
+}
+
+/// A condition to wait for.
+///
+/// This is implemented by different types directly, like [`Locator`]. But you can also use
+/// custom logic, using the [`Condition`] and [`Predicate`] newtypes.
+pub trait WaitCondition<'a, T> {
```

Comment:
I don't think you need `'a` here at all actually. Just make `ready` generic over `<'a>` instead. That will also simplify your bounds in the methods above further.

Reply:
Did you try making this change? I think that alone will simplify a lot of your bounds.

Reply:
I tried it before, and I failed with the blanket implementation of `WaitCondition` for the Fn type.

However, that isn't really being used, as I switched to the `Condition` and `Predicate` newtypes anyway, one more reason why I would like to keep them.

So `'a` is dropped now, and it still works with the blanket implementation dropped.